src/pstl_offload.cpp

```diff
@@ -96,59 +97,311 @@ __original_msize(void* __user_ptr)
-                __delayed_free = __h;
-            }
-            else
-            {
-                __original_free(__user_ptr);
+                throw std::bad_alloc();
             }
+            __delayed_free_list* __h = new (__buf) __delayed_free_list{__delayed_free, __user_ptr};
+            __delayed_free = __h;
+        }
+        else
+        {
+            __original_free(__user_ptr);
         }
     }
 }
 
 static std::size_t
 __internal_msize(void* __user_ptr)
 {
-    std::size_t __res = 0;
-    if (__user_ptr != nullptr)
+    if (__user_ptr == nullptr)
+    {
+        return 0;
+    }
+
+    __block_header* __header = static_cast<__block_header*>(__user_ptr) - 1;
+
+    if (__same_memory_page(__user_ptr, __header) && __header->_M_uniq_const == __uniq_type_const)
     {
+        return __header->_M_requested_number_of_bytes;
+    }
+
+    std::optional<std::size_t> __size = __large_aligned_ptrs_map::__get_size(__large_aligned_ptrs, __user_ptr);
 
-        __block_header* __header = static_cast<__block_header*>(__user_ptr) - 1;
+    return __size.has_value() ? __size.value() : __original_msize(__user_ptr);
+}
 
-        if (__same_memory_page(__user_ptr, __header) && __header->_M_uniq_const == __uniq_type_const)
+static void*
+__realloc_allocate_shared(__sycl_device_shared_ptr __device_ptr, void* __user_ptr, std::size_t __old_size, std::size_t __new_size)
+{
+    void* __new_ptr = __allocate_shared_for_device(std::move(__device_ptr), __new_size, alignof(std::max_align_t));
+
+    if (__new_ptr != nullptr)
+    {
+        std::memcpy(__new_ptr, __user_ptr, std::min(__old_size, __new_size));
+    }
+    else
+    {
+        errno = ENOMEM;
+    }
+    return __new_ptr;
+}
+
+_PSTL_OFFLOAD_EXPORT void*
+__realloc_impl(void* __user_ptr, std::size_t __new_size)
+{
+    assert(__user_ptr != nullptr);
+
+    if (!__new_size)
+    {
+        __internal_free(__user_ptr);
+        return nullptr;
+    }
+
+    if (__block_header* __header = static_cast<__block_header*>(__user_ptr) - 1; 
+        __same_memory_page(__user_ptr, __header) && __header->_M_uniq_const == __uniq_type_const)
+    {
+        if (__header->_M_requested_number_of_bytes == __new_size)
         {
-            __res = __header->_M_requested_number_of_bytes;
+            return __user_ptr;
         }
-        else
+
+        void* __result = __realloc_allocate_shared(__header->_M_device, __user_ptr,
+                                                   __header->_M_requested_number_of_bytes, __new_size);
+        if (__result)
         {
-            __res = __original_msize(__user_ptr);
+            __header->__free();
         }
+        return __result;
+    }
+
+    if (std::optional<__large_aligned_ptrs_map::__ptr_desc>
+            __desc = __large_aligned_ptrs_map::__unregister_ptr(__large_aligned_ptrs, __user_ptr))
+    {
+        if (__desc->_M_requested_number_of_bytes == __new_size)
+        {
+            return __user_ptr;
+        }
```

Comment:
> First of all, I think the check should be >= instead of ==.

Why? I disliked interesting cases like "the object was aligned, but now it's just 8 bytes, but we keep it in the table".

> The second issue is if that branch would be reached, __user_ptr would be removed from the pointers map, but not released and the next call to free would result in "original" free call.

Thanks, fixed.